File: backend/src/main/java/com/meditracker/backend/controller/MedicationController.java
package com.meditracker.backend.controller;

import com.meditracker.backend.model.Medication;
import com.meditracker.backend.service.MedicationService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/medications")
@CrossOrigin(origins = {"http://localhost:5173", "http://localhost:3000"}, allowCredentials = "true")
public class MedicationController {
    
    @Autowired
    private MedicationService medicationService;
    
    @GetMapping
    public ResponseEntity<List<Medication>> getAllMedications(@RequestParam String userId) {
        List<Medication> medications = medicationService.getMedicationsByUserId(userId);
        return ResponseEntity.ok(medications);
    }
    
    @GetMapping("/{id}")
    public ResponseEntity<Medication> getMedicationById(@PathVariable String id, @RequestParam String userId) {
        Medication medication = medicationService.getMedicationById(id, userId);
        return ResponseEntity.ok(medication);
    }
    
    @PostMapping
    public ResponseEntity<Medication> createMedication(@Valid @RequestBody Medication medication) {
        System.out.println("Received medication: " + medication.getName() + ", time: " + medication.getTime());
        Medication createdMedication = medicationService.createMedication(medication);
        return ResponseEntity.status(HttpStatus.CREATED).body(createdMedication);
    }
    
    @PutMapping("/{id}")
    public ResponseEntity<Medication> updateMedication(
            @PathVariable String id, 
            @Valid @RequestBody Medication medication,
            @RequestParam String userId) {
        Medication updatedMedication = medicationService.updateMedication(id, medication, userId);
        return ResponseEntity.ok(updatedMedication);
    }
    
    @DeleteMapping("/{id}")
    public ResponseEntity<Map<String, String>> deleteMedication(@PathVariable String id, @RequestParam String userId) {
        medicationService.deleteMedication(id, userId);
        return ResponseEntity.ok(Map.of("message", "Medication deleted successfully"));
    }
    
    @GetMapping("/search")
    public ResponseEntity<List<Medication>> searchMedications(
            @RequestParam String userId,
            @RequestParam String query) {
        List<Medication> medications = medicationService.searchMedications(userId, query);
        return ResponseEntity.ok(medications);
    }
    
    @GetMapping("/low-stock")
    public ResponseEntity<List<Medication>> getLowStockMedications(
            @RequestParam String userId,
            @RequestParam(defaultValue = "7") int threshold) {
        List<Medication> medications = medicationService.getLowStockMedications(userId, threshold);
        return ResponseEntity.ok(medications);
    }
    
    @PatchMapping("/{id}/stock")
    public ResponseEntity<Medication> updateStock(
            @PathVariable String id,
            @RequestParam String userId,
            @RequestParam int stock) {
        Medication medication = medicationService.updateStock(id, userId, stock);
        return ResponseEntity.ok(medication);
    }
    
    @GetMapping("/stats")
    public ResponseEntity<Map<String, Object>> getMedicationStats(@RequestParam String userId) {
        Map<String, Object> stats = medicationService.getMedicationStats(userId);
        return ResponseEntity.ok(stats);
    }
}

File: backend-node/controllers/medicationController.js
import Medication from '../models/Medication.js';

export const getAllMedications = async (req, res) => {
  try {
    const medications = await Medication.find({ userId: req.query.userId });
    res.json(medications);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const getMedicationById = async (req, res) => {
  try {
    const medication = await Medication.findOne({ _id: req.params.id, userId: req.query.userId });
    if (!medication) return res.status(404).json({ message: 'Medication not found' });
    res.json(medication);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const createMedication = async (req, res) => {
  try {
    const medication = new Medication(req.body);
    const savedMedication = await medication.save();
    res.status(201).json(savedMedication);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

export const updateMedication = async (req, res) => {
  try {
    const medication = await Medication.findOneAndUpdate(
      { _id: req.params.id, userId: req.query.userId },
      req.body,
      { new: true }
    );
    if (!medication) return res.status(404).json({ message: 'Medication not found' });
    res.json(medication);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

export const deleteMedication = async (req, res) => {
  try {
    const medication = await Medication.findOneAndDelete({ _id: req.params.id, userId: req.query.userId });
    if (!medication) return res.status(404).json({ message: 'Medication not found' });
    res.json({ message: 'Medication deleted successfully' });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const searchMedications = async (req, res) => {
  try {
    const medications = await Medication.find({
      userId: req.query.userId,
      name: { $regex: req.query.query, $options: 'i' }
    });
    res.json(medications);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const getLowStockMedications = async (req, res) => {
  try {
    const threshold = parseInt(req.query.threshold) || 7;
    const medications = await Medication.find({
      userId: req.query.userId,
      stock: { $lte: threshold }
    });
    res.json(medications);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const updateStock = async (req, res) => {
  try {
    const medication = await Medication.findOneAndUpdate(
      { _id: req.params.id, userId: req.query.userId },
      { stock: req.query.stock },
      { new: true }
    );
    if (!medication) return res.status(404).json({ message: 'Medication not found' });
    res.json(medication);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

export const getMedicationStats = async (req, res) => {
  try {
    const medications = await Medication.find({ userId: req.query.userId });
    const lowStockMedications = medications.filter(m => m.stock <= 7);
    const totalStock = medications.reduce((sum, m) => sum + (m.stock || 0), 0);
    
    res.json({
      totalMedications: medications.length,
      lowStockCount: lowStockMedications.length,
      totalStock
    });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};
